ang/test/civicase/PanelQuery.spec.js

```diff
@@ -0,0 +1,512 @@
+    var NO_OF_RESULTS = 10;
+
+    beforeEach(module('civicase.templates', 'civicase', 'crmUtil'));
+
+    beforeEach(inject(function (_$compile_, _$q_, _$rootScope_, _crmApi_) {
+      $compile = _$compile_;
+      $q = _$q_;
+      $rootScope = _$rootScope_;
+      crmApi = _crmApi_;
+
+      $scope = $rootScope.$new();
+      mockedResults = _.times(NO_OF_RESULTS, function () {
+        return { id: _.random(1, 10) };
+      });
+
+      crmApi.and.returnValue($q.resolve({
+        get: { values: mockedResults },
+        count: NO_OF_RESULTS
+      }));
+    }));
+
+    describe('[query] attribute', function () {
+      beforeEach(function () {
+        $scope.queryData = { entity: 'Foo', params: { foo: 'foo' } };
+        compileDirective();
+      });
+
+      it('store its value in its scope', function () {
+        expect(isolatedScope.query).toBeDefined();
+        expect(isolatedScope.query).toEqual($scope.queryData);
+      });
+
+      describe('one-way binding', function () {
+        var originalSource;
+
+        beforeEach(function () {
+          originalSource = $scope.queryData;
+          isolatedScope.query = { baz: 'baz' };
+
+          $scope.$digest();
+        });
+
+        it('has a one-way binding on the [query-data] value', function () {
+          expect($scope.queryData).toEqual(originalSource);
+        });
+      });
+
+      describe('`query` object', function () {
+        describe('when it doesn\'t have the `entity` property', function () {
+          beforeEach(function () {
+            $scope.queryData = { params: {} };
+          });
+
+          it('sends an error message', function () {
+            expect(compileDirective).toThrowError(/entity/);
+          });
+        });
+
+        describe('when the `entity` property is an empty string', function () {
+          beforeEach(function () {
+            $scope.queryData = { entity: '', params: {} };
+          });
+
+          it('sends an error message', function () {
+            expect(compileDirective).toThrowError(/entity/);
+          });
+        });
+      });
+    });
+
+    describe('[handlers] attribute', function () {
+      describe('results handler', function () {
+        var resultsHandler;
+        var newResults = ['foo', 'bar', 'baz'];
+
+        beforeEach(function () {
+          resultsHandler = jasmine.createSpy().and.callFake(function (results) {
+            return newResults;
+          });
+          $scope.handlersData = { results: resultsHandler };
+
+          compileDirective();
+        });
+
+        it('is called', function () {
+          expect(resultsHandler).toHaveBeenCalled();
+        });
+
+        it('receives the full results list as an argument', function () {
+          var arg = resultsHandler.calls.argsFor(0)[0];
+
+          expect(arg.length).toBe(NO_OF_RESULTS);
+          expect(arg.every(function (item) {
+            return typeof item.id !== 'undefined';
+          })).toBe(true);
+        });
```

Comment:
why not just:

```js
it('is called with the full list of results', function () {
  expect(resultsHandler).toHaveBeenCalledWith(newResults);
});
```

Reply:
The comparison will fail, `resultsHandler` receives the "old" list of results, while `newResults` represent what will come out of the handler